core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala

```diff
@@ -396,38 +430,54 @@ object ShardingContainerPoolBalancer extends LoadBalancerProvider {
    * @return an invoker to schedule to or None of no invoker is available
    */
   @tailrec
-  def schedule(
-    maxConcurrent: Int,
-    fqn: FullyQualifiedEntityName,
-    invokers: IndexedSeq[InvokerHealth],
-    dispatched: IndexedSeq[NestedSemaphore[FullyQualifiedEntityName]],
-    slots: Int,
-    index: Int,
-    step: Int,
-    stepsDone: Int = 0)(implicit logging: Logging, transId: TransactionId): Option[(InvokerInstanceId, Boolean)] = {
+  def schedule(maxConcurrent: Int,
+               fqn: FullyQualifiedEntityName,
+               invokers: IndexedSeq[InvokerHealth],
+               dispatched: IndexedSeq[NestedSemaphore[FullyQualifiedEntityName]],
+               slots: Int,
+               index: Int,
+               step: Int,
+               stepsDone: Int = 0,
+               preferredInvokers: Seq[Int] = Seq.empty[Int])(
+    implicit logging: Logging,
+    transId: TransactionId): Option[(InvokerInstanceId, Boolean)] = {
     val numInvokers = invokers.size
 
     if (numInvokers > 0) {
-      val invoker = invokers(index)
-      //test this invoker - if this action supports concurrency, use the scheduleConcurrent function
-      if (invoker.status.isUsable && dispatched(invoker.id.toInt).tryAcquireConcurrent(fqn, maxConcurrent, slots)) {
-        Some(invoker.id, false)
+      if (preferredInvokers.nonEmpty) {
+        val preferredInvoker = preferredInvokers
+          .find(
+            invokerIndex =>
+              invokerIndex < invokers.size && invokers(invokerIndex).status.isUsable && dispatched(
+                invokers(invokerIndex).id.toInt).tryAcquireConcurrent(fqn, maxConcurrent, slots))
+          .map(invokerIndex => Some((invokers(invokerIndex).id, false)))
+        if (preferredInvoker.nonEmpty) {
+          preferredInvoker.get
+        } else {
+          schedule(maxConcurrent, fqn, invokers, dispatched, slots, index, step, stepsDone + 1)
+        }
```

Comment:
nit: `preferredInvoker.getOrElse(schedule)`

Reply:
I had that originally, it surprisingly doesn't compile because of the way `@tailrec` works